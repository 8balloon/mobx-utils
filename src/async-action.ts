import {runInAction, action} from "mobx";
import {invariant} from "./utils";

// TODO: move async-action to mobx-core?

// method decorator:
export function asyncAction(target: Object, propertyKey: string, descriptor: PropertyDescriptor): PropertyDescriptor;

// non-decorator forms
export function asyncAction<R>(generator: () => IterableIterator<any>): () => Promise<R>;
export function asyncAction<A1>(generator: (a1: A1) => IterableIterator<any>): (a1: A1) => Promise<any>; // Ideally we want to have R instead of Any, but cannot specify R without specifying A1 etc... 'any' as result is better then not specifying request args
export function asyncAction<A1, A2, A3, A4, A5, A6, A7, A8>(generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => Promise<any>;
export function asyncAction<A1, A2, A3, A4, A5, A6, A7>(generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => Promise<any>;
export function asyncAction<A1, A2, A3, A4, A5, A6>(generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => Promise<any>;
export function asyncAction<A1, A2, A3, A4, A5>(generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => Promise<any>;
export function asyncAction<A1, A2, A3, A4>(generator: (a1: A1, a2: A2, a3: A3, a4: A4) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4) => Promise<any>;
export function asyncAction<A1, A2, A3>(generator: (a1: A1, a2: A2, a3: A3) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3) => Promise<any>;
export function asyncAction<A1, A2>(generator: (a1: A1, a2: A2) => IterableIterator<any>): (a1: A1, a2: A2) => Promise<any>;
export function asyncAction<A1>(generator: (a1: A1) => IterableIterator<any>): (a1: A1) => Promise<any>;
// ... with name
export function asyncAction<R>(name: string, generator: () => IterableIterator<any>): () => Promise<R>;
export function asyncAction<A1>(name: string, generator: (a1: A1) => IterableIterator<any>): (a1: A1) => Promise<any>; // Ideally we want to have R instead of Any, but cannot specify R without specifying A1 etc... 'any' as result is better then not specifying request args
export function asyncAction<A1, A2, A3, A4, A5, A6, A7, A8>(name: string, generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7, a8: A8) => Promise<any>;
export function asyncAction<A1, A2, A3, A4, A5, A6, A7>(name: string, generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6, a7: A7) => Promise<any>;
export function asyncAction<A1, A2, A3, A4, A5, A6>(name: string, generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5, a6: A6) => Promise<any>;
export function asyncAction<A1, A2, A3, A4, A5>(name: string, generator: (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5) => Promise<any>;
export function asyncAction<A1, A2, A3, A4>(name: string, generator: (a1: A1, a2: A2, a3: A3, a4: A4) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3, a4: A4) => Promise<any>;
export function asyncAction<A1, A2, A3>(name: string, generator: (a1: A1, a2: A2, a3: A3) => IterableIterator<any>): (a1: A1, a2: A2, a3: A3) => Promise<any>;
export function asyncAction<A1, A2>(name: string, generator: (a1: A1, a2: A2) => IterableIterator<any>): (a1: A1, a2: A2) => Promise<any>;
export function asyncAction<A1>(name: string, generator: (a1: A1) => IterableIterator<any>): (a1: A1) => Promise<any>;

export function asyncAction(arg1: any, arg2?: any): any {
    // decorator
    if (typeof arguments[1] === "string") {
        const name = arguments[1];
        const descriptor: PropertyDescriptor = arguments[2];
        if (descriptor && descriptor.value) {
            return Object.assign({}, descriptor, { value: createAsyncActionGenerator(name, descriptor.value) });
        } else {
            return Object.assign({}, descriptor, { set(v: any) {
                Object.defineProperty(this, name, { ...descriptor, value: asyncAction(name, v) });
            }});
        }
    }

    // direct invocation
    const generator = typeof arg1 === "string" ? arg2 : arg1;
    const name = typeof arg1 === "string" ? arg1 : (generator.name || "<unnamed async action>");

    invariant(typeof generator === "function", "asyncAction expects function as first arg, got: " + generator);
    return createAsyncActionGenerator(name, generator);
}

let generatorId = 0;

export function createAsyncActionGenerator(name: string, generator: Function) {
    // Implementation based on https://github.com/tj/co/blob/master/index.js
    return function() {
        const ctx = this;
        const args = arguments;
        return new Promise(function(resolve, reject) {
            const runId = ++generatorId;
            let stepId = 0;
            const gen = action(`${name} - runid: ${runId} - init`, generator).apply(ctx, args);
            onFulfilled(undefined); // kick off the process

            function onFulfilled(res: any) {
                let ret;
                try {
                    ret = action(`${name} - runid: ${runId} - yield ${stepId}`, gen.next).call(gen, res);
                } catch (e) {
                    return reject(e);
                }
                next(ret);
                return null;
            }

            function onRejected(err: any) {
                let ret;
                try {
                    ret = action(`${name} - runid: ${runId} - yield ${stepId}`, gen.throw).call(gen, err);
                } catch (e) {
                    return reject(e);
                }
                next(ret);
            }

            function next(ret: any) {
                if (ret.done) return resolve(ret.value);
                // TODO: support more type of values? See https://github.com/tj/co/blob/249bbdc72da24ae44076afd716349d2089b31c4c/index.js#L100
                invariant(ret.value && typeof ret.value.then === "function", "Only promises can be yielded to asyncAction, got: " + ret);
                return ret.value.then(onFulfilled, onRejected);
            }
        });
    };
}
