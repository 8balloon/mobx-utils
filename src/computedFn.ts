import { DeepMap } from "./deepMap"
import { IComputedValue, computed, onBecomeUnobserved } from "mobx";

export function computedFn<T extends Function>(fn: T, autoCleanup = true) {
  const d = new DeepMap<IComputedValue<any>>()

  return function(...args: any[]) {
    const self = this
    const entry = d.entry(args)
    if (entry.exists())
      return entry.get().get()

    // create new entry
    const c = computed(() => {
      return fn.apply(self, args)
    })
    if (!autoCleanup) onBecomeUnobserved(c, () => {
      d.entry(args).delete()
    })
    entry.set(c)
    return c.get()
  }
}