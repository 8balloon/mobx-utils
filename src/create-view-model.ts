import {observable, action, ObservableMap, asMap, isObservableObject} from "mobx";
import {invariant} from "./utils";

export interface IViewModel<T> {
    model: T;
    reset(): void;
    submit(): void;
    isDirty: boolean;
    isPropertyDirty(key: string): boolean;
}

const RESERVED_NAMES = ["model", "reset", "submit", "isDirty", "isPropertyDirty"];

class ViewModel<T> implements IViewModel<T> {
    @observable isDirty = false;
    localValues: ObservableMap<any> = asMap({});
    dirtyMap: ObservableMap<any> = asMap({});

    constructor(public model: T) {
        invariant(isObservableObject(model), "createViewModel expects an observable object");
        Object.keys(model).forEach(key => {
            invariant(RESERVED_NAMES.indexOf(key) === -1, `The propertyname ${key} is reserved and cannot be used with viewModels`);
            // TODO: create shallow clones for arrays and maps and observe those
            Object.defineProperty(this, key, {
                enumerable: true,
                configurable: true,
                get: () => {
                    if (this.isPropertyDirty(key))
                        return this.localValues.get(key);
                    else
                        return (this.model as any)[key];
                },
                set: action((value: any) => {
                    if (this.isPropertyDirty(key) || value !== (this.model as any)[key]) {
                        this.isDirty = true;
                        this.dirtyMap.set(key, true);
                        this.localValues.set(key, value);
                    }
                })
            });
        });
    }

    isPropertyDirty(key: string): boolean {
        return this.dirtyMap.get(key) === true;
    }

    @action submit() {
        if (this.isDirty) {
            this.isDirty = false;
            this.dirtyMap.entries().forEach(([key, dirty]) => {
                if (dirty === true) {
                    (this.model as any)[key] = this.localValues.get(key);
                    this.dirtyMap.set(key, false);
                    this.localValues.delete(key);
                }
            });
        }
    }

    @action reset() {
        if (this.isDirty) {
            this.isDirty = false;
            this.dirtyMap.entries().forEach(([key, dirty]) => {
                if (dirty === true) {
                    this.dirtyMap.set(key, false);
                    this.localValues.delete(key);
                }
            });
        }
    }
}

/**
 * `createViewModel` takes an object with observable properties (model)
 * and wraps a view model around it. The view model proxies all enumerable property of the original model with the following behavior:
 *  - as long as no new value has been assigned to the viewmodel property, the original property will be returned, and any future change in the model will be visible in the view model as well
 *  - once a new value has been assigned to a property of the viewmodel, that value will be returned during a read of that property in the future. However, the original model remain untouched until `submit()` is called.
 *
 * The viewmodel exposes the following additional methods, besides all the enumerable properties of the model:
 * - `submit()`: copies all the values of the viewmodel to the model and resets the state
 * - `reset()`: resets the state of the view model, abandoning all local modificatoins
 * - `isDirty`: observable property indicating if the viewModel contains any modifications
 * - `isPropertyDirty(propName)`: returns true if the specified property is dirty
 * - `model`: The original model object for which this viewModel was created
 *
 * N.B. doesn't support observable arrays and maps yet
 *
 * @example
 * class Todo {
 *   \@observable title = "Test"
 * }
 *
 * const model = new Todo()
 * const viewModel = createViewModel(model);
 *
 * autorun(() => console.log(viewModel.model.title, ",", viewModel.title))
 * // prints "Test, Test"
 * model.title = "Get coffee"
 * // prints "Get coffee, Get coffee", viewModel just proxies to model
 * viewModel.title = "Get tea"
 * // prints "Get coffee, Get tea", viewModel's title is now dirty, and the local value will be printed
 * viewModel.submit()
 * // prints "Get tea, Get tea", changes submitted from the viewModel to the model, viewModel is proxying again
 * viewModel.title = "Get cookie"
 * // prints "Get tea, Get cookie" // viewModel has diverged again
 * viewModel.reset()
 * // prints "Get tea, Get tea", changes of the viewModel have been abandoned
 *
 * @param {T} model
 * @returns {(T & IViewModel<T>)}
 * ```
 */
export function createViewModel<T>(model: T): T & IViewModel<T> {
    return new ViewModel(model) as any;
}