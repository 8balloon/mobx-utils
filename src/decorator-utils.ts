import { addHiddenProp } from "./utils"

type BabelDescriptor = PropertyDescriptor & { initializer?: () => any }

export function decorateMethodOrField(
    decoratorName: string,
    decorateFn: (pname: string, v: any) => any,
    target: object,
    prop: string,
    descriptor?: BabelDescriptor
) {
    if (descriptor) {
        return decorateMethod(decoratorName, decorateFn, prop, descriptor)
    } else {
        decorateField(decorateFn, target, prop)
    }
}

export function decorateMethod(
    decoratorName: string,
    decorateFn: (pname: string, v: any) => any,
    prop: string,
    descriptor: BabelDescriptor
) {
    if (descriptor.get !== undefined) {
        return fail(`${decoratorName} cannot be used with getters`)
    }

    // babel / typescript
    // @action method() { }
    if (descriptor.value) {
        // typescript
        return {
            value: decorateFn(prop, descriptor.value),
            enumerable: false,
            configurable: true, // See #1477
            writable: true, // for typescript, this must be writable, otherwise it cannot inherit :/ (see inheritable actions test)
        }
    }

    // babel only: @action method = () => {}
    const { initializer } = descriptor
    return {
        enumerable: false,
        configurable: true, // See #1477
        writable: true, // See #1398
        initializer() {
            // N.B: we can't immediately invoke initializer; this would be wrong
            return decorateFn(prop, initializer!.call(this))
        },
    }
}

export function decorateField(
    decorateFn: (pname: string, v: any) => any,
    target: object,
    prop: string
) {
    // Simple property that writes on first invocation to the current instance
    Object.defineProperty(target, prop, {
        configurable: true,
        enumerable: false,
        get() {
            return undefined
        },
        set(value) {
            addHiddenProp(this, prop, decorateFn(prop, value))
        },
    })
}
