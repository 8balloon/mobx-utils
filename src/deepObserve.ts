import {
    observe,
    isObservableMap,
    isObservableObject,
    isObservableArray,
    IObjectDidChange,
    IArrayChange,
    IArraySplice,
    IMapDidChange,
    values,
    entries
} from "mobx"
import { IDisposer } from "./utils"

type IChange = IObjectDidChange | IArrayChange | IArraySplice | IMapDidChange

type Entry = {
    dispose: IDisposer
    path: string
    parent?: Entry
}

function buildPath(entry: Entry): string {
    const res: string[] = []
    while (entry.parent) {
        res.push(entry.path)
        entry = entry.parent
    }
    return res.reverse().join("/")
}

export function deepObserve(
    target: any,
    listener: (change: IChange, path: string) => void
): IDisposer {
    const entrySet = new WeakMap<any, Entry>()

    function genericListener(change: IChange) {
        const entry = entrySet.get(change.object)
        processChange(change, entry)
        listener(change, buildPath(entry))
    }

    function processChange(change: IChange, parent: Entry) {
        switch(change.type) {
            // Object changes
            case "add": // also for map
                observeRecursively(change.newValue, parent, change.name)
                break
            case "update": // also for array and map
                unobserveRecursively(change.oldValue)
                observeRecursively(change.newValue, parent, (change as any).name || ("" + (change as any).index))
                break
            case "remove": // object
            case "delete": // map
                unobserveRecursively(change.oldValue)
                break
            // Array changes
            case "splice":
                change.removed.map(unobserveRecursively)
                change.added.forEach((value, idx) => observeRecursively(value, parent, "" + (change.index + idx)))
                // update paths
                for (let i = change.index + change.addedCount; i < change.object.length; i++) {
                    const entry = entrySet.get(change.object[i])
                    if (entry) entry.path = "" + i
                }
                break
        }
    }

    function observeRecursively(target: any, parent: Entry, path: string) {
        if (isObservableObject(target) || isObservableArray(target) || isObservableMap(target)) {
            if (entrySet.has(target)) {
                const entry = entrySet.get(target)
                if (entry.parent !== parent || entry.path !== path)
                    throw new Error(
                        `The same observable object cannot appear twice in the same tree, trying to assign it to '${buildPath(parent)}/${path}', but it already exists at '${buildPath(entry.parent)}/${entry.path}'`
                    )
            } else {
                const entry = {
                    parent,
                    path,
                    dispose: observe(target, genericListener)
                }
                entrySet.set(target, entry)
                entries(target).forEach(([key, value]) => observeRecursively(value, entry, key))
            }
        }
    }

    function unobserveRecursively(target: any) {
        const entry = entrySet.get(target)
        if (!entry) return
        entrySet.delete(target)
        entry.dispose()
        values(target).forEach(unobserveRecursively)
    }

    observeRecursively(target, undefined, "")

    return () => {
        unobserveRecursively(target)
    }
}
