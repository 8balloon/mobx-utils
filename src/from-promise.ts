import {IObservableValue, observable, action} from "mobx";
import {IDENTITY, IThenable} from "./utils";

export type PromiseState = "pending" | "fulfilled" | "rejected";

export interface IPromiseBasedObservable<T> {
    value: T;
    state: PromiseState;
    reason: any;
    promise: IThenable<T>;
}

class PromiseBasedObservable<T> implements IPromiseBasedObservable<T> {
    private _observable: IObservableValue<T>;
    private _state: IObservableValue<PromiseState> = observable("pending" as PromiseState);
    private _reason: IObservableValue<any> = observable(undefined as any);

    constructor(public promise: IThenable<T>, initialValue: T = undefined, private modifier = IDENTITY) {
        this._observable = observable(modifier(initialValue));
        promise.then(action("observableFromPromise-resolve", (value: T) => {
            this._observable.set(this.modifier(value));
            this._state.set("fulfilled");
        })).catch(action("observableFromPromise-reject", (reason: any) => {
            this._reason.set(reason);
            this._state.set("rejected");
        }));
    }

    get value(): T {
        return this._observable.get();
    }
    get state(): PromiseState {
        return this._state.get();
    }
    get reason(): any {
        return this._reason.get();
    }
}

/**
 * `fromPromise` takes a Promise and returns an object with 3 observable properties that track
 * the status of the promise. The returned object has the following observable properties:
 *  - `value`: either the initial value, or the value the Promise resolved to
 *  - `state`: one of `"pending"`, `"fulfilled"` or `"rejected"`
 *  - `reason`: the reject reason if the state is `"rejected"`
 *  - `promise`: (not observable) the original promise object
 *
 * @example
 * const fetchResult = fromPromise(fetch("http://someurl"))
 *
 * // combine with when..
 * when(
 *   () => fetchResult.state !== "pending"
 *   () => {
 *     console.log("Got ", fetchResult.reason || fetchResult.value)
 *   }
 * )
 *
 * // or a mobx-react component..
 * const myComponent = observer(({ fetchResult }) => {
 *   switch(fetchResult.state) {
 *      case "pending": return <div>Loading...</div>
 *      case "rejected": return <div>Ooops... {fetchResult.reason}</div>
 *      case "fulfilled": return <div>Gotcha: {fetchResult.value}</div>
 *   }
 * })
 *
 * @param {IThenable<T>} promise The promise which will be observed
 * @param {T} [initialValue=undefined] Optional predefined initial value
 * @param {any} [modifier=IDENTITY] MobX modifier, e.g. `asFlat`, to be applied to the resolved value
 * @returns {IPromiseBasedObservable<T>}
 */
export function fromPromise<T>(promise: IThenable<T>, initialValue: T = undefined, modifier =  IDENTITY): IPromiseBasedObservable<T> {
    return new PromiseBasedObservable(promise, initialValue, modifier);
}

