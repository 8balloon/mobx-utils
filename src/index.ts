export * from "./from-promise";
export * from "./guarded-when";
export * from "./keep-alive";
export * from "./queue-processor";
export * from "./lazy-observable";
export * from "./subscribing-observable";
export * from "./utils";