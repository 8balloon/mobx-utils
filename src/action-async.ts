import { _startAction, _endAction, IActionRunInfo } from "mobx"
import { invariant } from "./utils"
import { decorateMethodOrField } from "./decorator-utils"
import { fail } from "./utils"

let runId = 0
const unfinishedIds = new Set<number>()
const currentlyActiveIds = new Set<number>()

interface IActionAsyncContext {
    runId: number
    step: number
    actionRunInfo: IActionRunInfo
    actionName: string
    scope: any
    args: IArguments
}

let taskOrderPromise = Promise.resolve()
const emptyFunction = () => Promise.resolve()

const actionAsyncContextStack: IActionAsyncContext[] = []

export async function task<R>(value: R | PromiseLike<R>): Promise<R> {
    const ctx = actionAsyncContextStack[actionAsyncContextStack.length - 1]

    if (!ctx) {
        fail(
            "'actionAsync' context not present when running 'task'. did you await inside an 'actionAsync' without using 'task(promise)'? did you forget to await the task?"
        )
    }

    const { runId, actionName, args, scope, actionRunInfo, step } = ctx
    const nextStep = step + 1
    actionAsyncContextStack.pop()
    _endAction(actionRunInfo)
    currentlyActiveIds.delete(runId)

    try {
        const ret = await value

        // we use this trick to force a proper order of execution
        // even for immediately resolved promises
        taskOrderPromise = taskOrderPromise.then(emptyFunction)
        await taskOrderPromise

        return ret
    } finally {
        // only restart if it not a dangling promise (the action is not yet finished)
        if (unfinishedIds.has(runId)) {
            const actionRunInfo = _startAction(
                getActionAsyncName(actionName, runId, nextStep),
                this,
                args
            )

            actionAsyncContextStack.push({
                runId,
                step: nextStep,
                actionRunInfo,
                actionName,
                args,
                scope
            })
            currentlyActiveIds.add(runId)
        }
    }
}

// method decorator
export function actionAsync(
    target: object,
    propertyKey: string,
    descriptor: PropertyDescriptor
): PropertyDescriptor

// field decorator
export function actionAsync(target: object, propertyKey: string): void

// non-decorator forms
export function actionAsync<F extends (...args: any[]) => Promise<any>>(name: string, fn: F): F
export function actionAsync<F extends (...args: any[]) => Promise<any>>(fn: F): F

// base

/**
 * Alternative syntax for async actions, similar to `flow` but more compatible with
 * Typescript typings. Not to be confused with `asyncAction`, which is deprecated.
 *
 * `actionAsync` can be used either as a decorator or as a function.
 * It takes an async function that internally must use `await task(promise)` rather than
 * the standard `await promise`.
 *
 * When using the mobx devTools, an asyncAction will emit `action` events with names like:
 * * `"fetchUsers - runid 6 - step 0"`
 * * `"fetchUsers - runid 6 - step 1"`
 * * `"fetchUsers - runid 6 - step 2"`
 *
 * The `runId` represents the action instance. In other words, if `fetchUsers` is invoked
 * multiple times concurrently, the events with the same `runid` belong together.
 * The `step` number indicates the code block that is now being executed.
 *
 * @example
 * import {actionAsync, task} from "mobx-utils"
 *
 * let users = []
 *
 * const fetchUsers = actionAsync("fetchUsers", async (url) => {
 *   const start = Date.now()
 *   // note the use of task when awaiting!
 *   const data = await task(window.fetch(url))
 *   users = await task(data.json())
 *   return start - Date.now()
 * })
 *
 * const time = await fetchUsers("http://users.com")
 * console.log("Got users", users, "in ", time, "ms")
 *
 * @example
 * import {actionAsync, task} from "mobx-utils"
 *
 * mobx.configure({ enforceActions: "observed" }) // don't allow state modifications outside actions
 *
 * class Store {
 *   \@observable githubProjects = []
 *   \@observable = "pending" // "pending" / "done" / "error"
 *
 *   \@actionAsync
 *   async fetchProjects() {
 *     this.githubProjects = []
 *     this.state = "pending"
 *     try {
 *       // note the use of task when awaiting!
 *       const projects = await task(fetchGithubProjectsSomehow())
 *       const filteredProjects = somePreprocessing(projects)
 *       // the asynchronous blocks will automatically be wrapped actions
 *       this.state = "done"
 *       this.githubProjects = filteredProjects
 *     } catch (error) {
 *        this.state = "error"
 *     }
 *   }
 * }
 */
export function actionAsync(arg1?: any, arg2?: any, arg3?: any): any {
    // decorator
    if (typeof arguments[1] === "string") {
        return decorateMethodOrField(
            "@actionAsync",
            (prop, v) => {
                return actionAsyncFn(prop, v)
            },
            arg1,
            arg2,
            arg3
        )
    }

    // direct invocation
    const actionName = typeof arg1 === "string" ? arg1 : arg1.name || "<unnamed action>"
    const fn = typeof arg1 === "function" ? arg1 : arg2

    return actionAsyncFn(actionName, fn)
}

function actionAsyncFn(actionName: string, fn: Function): Function {
    if (!_startAction || !_endAction) {
        fail("'actionAsync' requires mobx >=5.13.1 or >=4.13.1")
    }

    invariant(typeof fn === "function", "'asyncAction' expects a function")
    if (typeof actionName !== "string" || !actionName)
        fail(`actions should have valid names, got: '${actionName}'`)

    return async function(this: any, ...args: any) {
        const nextRunId = runId++
        unfinishedIds.add(nextRunId)

        const actionRunInfo = _startAction(getActionAsyncName(actionName, nextRunId, 0), this, args)

        actionAsyncContextStack.push({
            runId: nextRunId,
            step: 0,
            actionRunInfo,
            actionName,
            args,
            scope: this
        })
        currentlyActiveIds.add(nextRunId)

        const finish = (err: any) => {
            unfinishedIds.delete(nextRunId)

            const ctx = actionAsyncContextStack.pop()
            if (!ctx || ctx.runId !== nextRunId) {
                // push it back if invalid
                if (ctx) {
                    actionAsyncContextStack.push(ctx)
                }

                let msg = `invalid 'actionAsync' context when finishing action '${actionName}'.`
                if (!ctx) {
                    msg += " no action context could be found instead."
                } else {
                    msg += ` an action context for '${ctx.actionName}' was found instead.`
                }
                msg +=
                    " did you await inside an 'actionAsync' without using 'task(promise)'? did you forget to await the task?"
                fail(msg)
            }
            ctx.actionRunInfo.error = err
            _endAction(ctx.actionRunInfo)
            currentlyActiveIds.delete(nextRunId)

            if (err) {
                throw err
            }
        }

        let promise: any
        try {
            promise = fn.apply(this, args)
        } catch (err) {
            finish(err)
        }

        // are we done sync? (no task run)
        if (currentlyActiveIds.has(nextRunId)) {
            finish(undefined)
            return promise
        }

        let ret: any
        try {
            ret = await promise
        } catch (err) {
            finish(err)
        }
        finish(undefined)
        return ret
    }
}

function getActionAsyncName(actionName: string, runId: number, step: number) {
    return `${actionName} - runid ${runId} - step ${step}`
}
