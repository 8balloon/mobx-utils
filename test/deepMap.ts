import { DeepMap } from "../src/deepMap"

test("args length 2", () => {
    const d = new DeepMap<number>()

    const e = d.get(["hello", "world"])

    expect(e.exists()).toBe(false)

    expect(() => {
        e.get()
    }).toThrowErrorMatchingSnapshot()

    e.set(3)

    expect(e.get()).toBe(3)
    expect(e.exists()).toBe(true)

    const e2 = d.get(["hello", "world"])
    expect(e2.exists()).toBe(true)
    expect(e2.get()).toBe(3)

    e2.set(4)
    expect(() => {
        e.get()
    }).toThrowErrorMatchingSnapshot()

    expect(d.get(["hello", "world"]).get()).toBe(4)

    expect(() => d.get(["bla"])).toThrowErrorMatchingSnapshot()

    d.get(["coffee", "tea"]).set(100)

    const e3 = d.get(["hello", "universe"])
    e3.set(42)
    expect(e3.exists()).toBe(true)
    expect(e3.get()).toBe(42)

    expect(d.get(["hello", "world"]).get()).toBe(4)
    expect(d.get(["hello", "universe"]).get()).toBe(42)
    expect(d.get(["coffee", "tea"]).get()).toBe(100)

    d.get(["hello", "world"]).delete()
    expect(d.get(["hello", "world"]).exists()).toBe(false)
    expect(d.get(["hello", "universe"]).get()).toBe(42)
    
    d.get(["coffee", "tea"]).delete()
    expect((d as any).store.size).toBe(1)

    expect(d.get(["hello", "universe"]).get()).toBe(42)
    d.get(["hello", "universe"]).delete()
    expect((d as any).store.size).toBe(0)
})

test("really deep", () => {
  const d = new DeepMap<number>()
  const path = ["a", "b", "c", "d", "e"]
  expect(d.get(path).exists()).toBe(false)
  d.get(path).set(3)
  expect(d.get(path).exists()).toBe(true)
  expect(d.get(path).get()).toBe(3)
  d.get(path).set(4)
  expect(d.get(path).get()).toBe(4)

  d.get(path).delete()
  expect((d as any).store.size).toBe(0)
})

test("really shallow", () => {
  const d = new DeepMap<number>()
  const path = []
  expect(d.get(path).exists()).toBe(false)
  d.get(path).set(3)
  expect(d.get(path).exists()).toBe(true)
  expect(d.get(path).get()).toBe(3)
  d.get(path).set(4)
  expect(d.get(path).get()).toBe(4)

  d.get(path).delete()
  expect((d as any).store.size).toBe(0)
})
