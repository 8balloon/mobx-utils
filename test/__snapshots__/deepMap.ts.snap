// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`args length 2 1`] = `"Entry doesn't exist"`;

exports[`args length 2 2`] = `"Concurrent modification exception"`;

exports[`args length 2 3`] = `"DeepMap should be used with functions with a consistent length, expected: 2, got: 1"`;
